src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/LogicalOpUtils.java

```diff
@@ -186,6 +179,41 @@ else if ( req instanceof BGPRequest) {
                 return new HashSet<>( bgp.getTriplePatterns() );
             }
         }
+        else if( req instanceof SPARQLRequest ) {
+            final SPARQLGraphPattern graphPattern = ((SPARQLRequest) req).getQueryPattern();
+
+            final Set<TriplePattern> tps = new HashSet<>();
+            for ( final SPARQLGraphPattern pattern: ( (SPARQLGroupPatternImpl) graphPattern).getSubPatterns() ) {
```

Comment:
It is not guaranteed that `graphPattern` is of type `SPARQLGroupPatternImpl`. I am afraid that you have to consider all the various subinterfaces and the classes that implement `SPARQLGraphPattern` (e.g., `TriplePattern`, `BGP`, `SPARQLGroupPattern`, `GenericSPARQLGraphPatternImpl1`, ...)